client/session.go

```diff
@@ -1541,7 +1577,137 @@ func (s *session) streamBlocksMetadataFromPeer(
 	}
 
 	for moreResults {
-		if err := retrier.Attempt(fetchFn); err != nil {
+		if err := s.streamBlocksRetrier.Attempt(fetchFn); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// streamBlocksMetadataFromPeerV2 has several heap allocated anonymous
+// function, however, they're only allocated once per peer/shard combination
+// for the entire peer bootstrapping process so performance is acceptable
+func (s *session) streamBlocksMetadataFromPeerV2(
+	namespace ts.ID,
+	shard uint32,
+	peer peer,
+	start, end time.Time,
+	metadataCh chan<- blocksMetadata,
+	progress *streamFromPeersMetrics,
+) error {
+	var (
+		pageToken              []byte
+		optionIncludeSizes     = true
+		optionIncludeChecksums = true
+		optionIncludeLastRead  = true
+		moreResults            = true
+	)
+
+	// Declare before loop to avoid redeclaring each iteration
+	attemptFn := func(client rpc.TChanNode) error {
+		tctx, _ := thrift.NewContext(s.streamBlocksMetadataBatchTimeout)
+		req := rpc.NewFetchBlocksMetadataRawV2Request()
+		req.NameSpace = namespace.Data().Get()
+		req.Shard = int32(shard)
+		req.RangeStart = start.UnixNano()
+		req.RangeEnd = end.UnixNano()
+		req.Limit = int64(s.streamBlocksBatchSize)
+		req.PageToken = pageToken
+		req.IncludeSizes = &optionIncludeSizes
+		req.IncludeChecksums = &optionIncludeChecksums
+		req.IncludeLastRead = &optionIncludeLastRead
+
+		progress.metadataFetchBatchCall.Inc(1)
+		result, err := client.FetchBlocksMetadataRawV2(tctx, req)
+		if err != nil {
+			progress.metadataFetchBatchError.Inc(1)
+			return err
+		}
+
+		progress.metadataFetchBatchSuccess.Inc(1)
+		progress.metadataReceived.Inc(int64(len(result.Elements)))
+
+		if result.NextPageToken != nil {
+			// copy will not extend the destination slice automatically
+			nextPageTokenLen := len(result.NextPageToken)
+			pageTokenLen := len(pageToken)
+			delta := nextPageTokenLen - pageTokenLen
+			if delta > 0 {
+				for i := 0; i < delta; i++ {
+					pageToken = append(pageToken, '\x00')
+				}
+			}
+			// Copy the pageToken so we don't have to keep the result around
+			copy(pageToken, result.NextPageToken)
+			// Make sure we don't include extraneous bytes if the new page token is
+			// shorter than the previous one (for whateve reason)
+			pageToken = pageToken[:nextPageTokenLen]
```

Comment:
Hey instead of all this code, better to just do:

Reply:
that...is much better haha thanks